From 8e75d9e4f559b2d504e07443348bfeb44fcd397a Mon Sep 17 00:00:00 2001
From: Xiaowen Liu <b37945@freescale.com>
Date: Mon, 8 Apr 2013 16:02:47 +0800
Subject: [PATCH 445/486] ENGR00215174 fix cts verifier NV21 color showing
 incorrect.

add convertNV12toNV21 color conversion function except 1080p resolution.
because 1080p can't achieve 30fps to do the conversion.

Signed-off-by: Xiaowen Liu <b37945@freescale.com>
---
 mx6/libcamera2/StreamAdapter.cpp | 34 ++++++++++++++++++++++++++++++++++
 mx6/libcamera2/StreamAdapter.h   |  1 +
 2 files changed, 35 insertions(+)

diff --git a/mx6/libcamera2/StreamAdapter.cpp b/mx6/libcamera2/StreamAdapter.cpp
index bd4f361..a01a6b3 100755
--- a/mx6/libcamera2/StreamAdapter.cpp
+++ b/mx6/libcamera2/StreamAdapter.cpp
@@ -285,6 +285,36 @@ void StreamAdapter::convertNV12toYV12(StreamBuffer* dst, StreamBuffer* src)
     }
 }
 
+void StreamAdapter::convertNV12toNV21(StreamBuffer* dst, StreamBuffer* src)
+{
+    int Ysize = 0, UVsize = 0;
+    uint8_t *srcIn, *dstOut;
+    uint32_t *UVout;
+    struct g2d_buf s_buf, d_buf;
+    int size = (src->mSize > dst->mSize) ? dst->mSize : src->mSize;
+
+    Ysize  = src->mWidth * src->mHeight;
+    UVsize = src->mWidth *  src->mHeight >> 2;
+    srcIn = (uint8_t *)src->mVirtAddr;
+    dstOut = (uint8_t *)dst->mVirtAddr;
+    UVout = (uint32_t *)(dstOut + Ysize);
+
+    if (g2dHandle != NULL) {
+        s_buf.buf_paddr = src->mPhyAddr;
+        s_buf.buf_vaddr = src->mVirtAddr;
+        d_buf.buf_paddr = dst->mPhyAddr;
+        d_buf.buf_vaddr = dst->mVirtAddr;
+        g2d_copy(g2dHandle, &d_buf, &s_buf, size);
+    }
+    else {
+        memcpy(dstOut, srcIn, size);
+    }
+
+    for (int k = 0; k < UVsize/2; k++) {
+        __asm volatile ("rev16 %0, %0" : "+r"(*UVout));
+        UVout += 1;
+    }
+}
 
 int StreamAdapter::processFrame(CameraFrame *frame)
 {
@@ -307,6 +337,10 @@ int StreamAdapter::processFrame(CameraFrame *frame)
             buffer.mFormat == HAL_PIXEL_FORMAT_YCbCr_420_P) {
         convertNV12toYV12(&buffer, frame);
     }
+    else if (mStreamId == STREAM_ID_PRVCB && buffer.mWidth <= 1280 &&
+            buffer.mFormat == HAL_PIXEL_FORMAT_YCbCr_420_SP) {
+        convertNV12toNV21(&buffer, frame);
+    }
     else if (g2dHandle != NULL) {
         struct g2d_buf s_buf, d_buf;
         s_buf.buf_paddr = frame->mPhyAddr;
diff --git a/mx6/libcamera2/StreamAdapter.h b/mx6/libcamera2/StreamAdapter.h
index 60d4de3..0efaa27 100755
--- a/mx6/libcamera2/StreamAdapter.h
+++ b/mx6/libcamera2/StreamAdapter.h
@@ -65,6 +65,7 @@ public:
     void setErrorListener(CameraErrorListener *listener);
     void showFps();
     void convertNV12toYV12(StreamBuffer* dst, StreamBuffer* src);
+    void convertNV12toNV21(StreamBuffer* dst, StreamBuffer* src);
 
     enum StreamCommands {
         STREAM_START,
-- 
1.8.0

